lib/system/web/src/keyboard_engine.rs

```diff
@@ -0,0 +1,122 @@
+    pub target:   web_sys::EventTarget
+}
+
+impl KeyboardEngine {
+    pub fn new(target: &web_sys::EventTarget) -> Result<Self> {
+        let state = Rc::new(RefCell::new(BTreeSet::new()));
+        let bindings = Rc::new(RefCell::new(HashMap::new()));
+
+        KeyboardEngine::init(target, Rc::clone(&bindings), Rc::clone(&state));
+
+        Ok(Self {
+            bindings,
+            state,
+            target: target.clone()
+        })
+    }
+
+    pub fn from_tag_name(html_tag: &str) -> Result<Self> {
+        let element = document()?
+            .get_elements_by_tag_name(html_tag)
+            .item(0).ok_or_else(|| Error::missing("there is no such element"));
+
+        KeyboardEngine::new(&element.unwrap())
+    }
+
+    fn init (
+        target:   &web_sys::EventTarget,
+        bindings: Rc<RefCell<Bindings>>,
+        state:    Rc<RefCell<BTreeSet<u32>>>
+    ) {
+        {
+            let state = Rc::clone(&state);
+            let bindings = Rc::clone(&bindings);
+            let keydown_callback = Box::new(move |event: web_sys::KeyboardEvent| {
+                if event.repeat() {
+                    return;
+                }
+
+                let key_code = event.key_code();
+                state.borrow_mut().insert(key_code);
+
+                let key = state.borrow();
+                if let Some(vec) = bindings.borrow_mut().get_mut(&key) {
+                    vec.iter_mut().for_each(|func| func());
+                }
+            });
+            let keydown_callback = Closure::wrap(
+                keydown_callback as Box<dyn FnMut(_)>);
+            target.add_event_listener_with_callback(
+                "keydown", keydown_callback.as_ref().unchecked_ref())
+                .unwrap();
+            keydown_callback.forget();
+        }
+
+        {
+            let keyup_callback = Box::new(move |event: KeyboardEvent| {
+                let key_code = event.key_code();
+                state.borrow_mut().remove(&key_code);
+
+                let key = state.borrow();
+                if let Some(vec) = bindings.borrow_mut().get_mut(&key) {
+                    vec.iter_mut().for_each(|func| func());
+                }
+            });
+            let keyup_callback = Closure::wrap(
+                keyup_callback as Box<dyn FnMut(_)>);
+            target.add_event_listener_with_callback(
+                "keyup", keyup_callback.as_ref().unchecked_ref())
+                .unwrap();
+            keyup_callback.forget();
+        }
+    }
+
+    /// Capture key combination
+    ///
+    /// Uses u32 as key code from WebIDL
+    /// https://github.com/rustwasm/wasm-bindgen/blob/913fdbc3daff65952b5678a34b98e07d4e6e4fbb/crates/web-sys/webidls/enabled/KeyEvent.webidl
+    pub fn capture(
+        & self,
+        combination: Vec<u32>,
+        callback:    Box<dyn FnMut()>
+    ) {
+        let key = combination.iter().copied().collect::<BTreeSet<u32>>();
+        let mut bindings = self.bindings.borrow_mut();
```

Comment:
I would create some kind of callback-abstraction, which hides all internal mutability and exposes simple API. Using `borrow_mut` in various functions is often considered a bad practice. Instead, we can create something like `BindingRegistry` (or other name), which handles all bindings for us and exposes a nice API :)

Reply:
The main problem that I have at this point that I need to
1. Have 2 closures with **static** lifetime which mutates the data
2. So I'm not able to borrow the data, only move i
3. And I cannot use `Rc` without `RefCell` or `Cell` cause I have multiple borrowing

If I got right, you suggest me to move bindins into `BindingRegistry`, but it doesn't change something, cause either I need to store internal data for `BindingRegistry` in `RefCell` or wrap it in `RefCell` itself as it's done right now. If I move everything in separated structures, I can make `KeyboardEngine` without internal mutability but I get new two that have it

How do you see I ca
1. Have 2 static closures with moved data that I should mutate
2. And don't use internal mutability at the same time
3. And what pattern at this point can do it nicer as it's right now?